### DISCUSSION TITLE
How to get value from multiple type possibilities

### DISCUSSION THREAD

**Discussion by sgleske-ias** [mikefarah/yq#773] [Q&A] [answered]
I have a YAML file which describes languages and toolchains (e.g. `.jervis.yml` or Travis CI `.travis.yml`).

# Goal

Get a version of a toolchain from a YAML file (e.g. python) when the toolchain could look like different types (or have nothing defined).

# YAML object type possibilities

Please note: I would look to the `python` key for the version of python.

`python` not defined.

```yaml
language: python
```

or `python` as a number

```yaml
language: python
python: 3.6
```

or `python` as a string

```yaml
language: python
python: "3.6"
```

or `python` as a seq

```yaml
language: python
python:
  - 3.6
  - 2.7
```

# Goal

Trying to get the version of python from a list (the first item only), or fall back to getting the value (e.g. `3.6`), or fall back to a documented default `2.7`.

# My hack

In bash, I could put in a series of defaults.  For example,

```bash
python_version="$(yq eval '.python[0]' .jervis.yml)"
if [ -z "${python_version:-}" -o "${python_version:-}" = null ]; then
    python_version="$(yq eval '.python // "2.7"' .jervis.yml)"
fi

echo "Python version found: ${python_version}"
```

The above script is the only way I could figure out how to make it work.  Is there a `yq` one liner which would work?  I'm at my wit's end on this one.

### My failed attempt at a one liner

Please note: I tried

```
yq eval '.python[0] // .python // "2.7"' .jervis.yml
```

But it breaks down so I must be misunderstanding it.  I also attempted some type checking but it didn't appear to work the way I thought it would.

```
yq eval '(.python | select((. | tag) == "!!seq") | .[0]) // .python // "2.7"' .jervis.yml
```

**Comment by mikefarah** [comment, marked as answer]
Yeah I see - let me take a look I think it's a problem with the `//` operator

**Reply by mikefarah** [reply]
Yep found the bug - will fix in the next release

**Reply by mikefarah** [reply]
FYI - with the new release, the following will work:

```
yq e '(.python | select(tag == "!!seq") | .[0]) // .python // "2.7"'
```

**Reply by sgleske-ias** [reply]
Great; thanks @mikefarah !  I thought it could be a bug but I wasn't trusting my instinct since I'm new to yq.

Edit for credit: https://github.com/mikefarah/yq/commit/25e0a824c58ae41b2fdfc2d08fa54804d7081005 nice

**Reply by mikefarah** [reply]
Hi @sgleske-ias this has been fixed in the latest release (4.7.0) let me know how it goes :)

**Reply by sgleske-ias** [reply]
@mikefarah Thanks! I confirmed the fix by downloading 4.7.1.
